File: vhdl_toolkit/vhdlConvertor/hdlObjects/Operator.java
package hdlObjects;

import java.util.List;
import java.util.Vector;

import org.json.JSONException;
import org.json.JSONObject;

public class Operator {
	Expr op0;
	OperatorType operator;
	Expr op1;
	List<Expr> operands;
	public Operator(Expr op0, OperatorType operatorType, Expr op1) {
		this.op0 = op0;
		this.operator = operatorType;
		this.op1 = op1;
	}
	public Operator(Expr op0, OperatorType operatorType, List<Expr> operands) {
		assert (operatorType == OperatorType.CALL);
		this.op0 = op0;
		this.operator = operatorType;
		this.operands = operands;

	}
	public JSONObject toJson() throws JSONException {
		JSONObject o = new JSONObject();
		o.put("op0", op0.toJson());
		o.put("operator", operator.toString());
		int arity = operator.arity();
		switch (arity) {
			case -1 :
				List<JSONObject> opList = new Vector<JSONObject>();
				for (Expr e : operands)
					opList.add(e.toJson());
				o.put("operands", opList);
				break;
			case 1 :
				break;
			case 2 :
				o.put("op1", op1.toJson());
				break;
			default :
				System.err.println(
						String.format("Invalid arity of operator %s  (%d)",
								operator.toString(), arity));
		}
		return o;
	}
}

File: hdl_toolkit/parser/hdlConvertor/src/hdlObjects/operator.cpp
#include "operator.h"

Operator::Operator() {
	operands = NULL;
	op0 = NULL;
	op1 = NULL;
	op = ARROW;
}

Operator::Operator(Expr* op0, OperatorType operatorType, Expr* op1) {
	this->op0 = op0;
	this->op = operatorType;
	this->op1 = op1;
	operands = NULL;
}
Operator * Operator::call(Expr* fn, std::vector<Expr*> * operands) {
	Operator * o = new Operator();
	o->op0 = fn;
	o->op = CALL;
	o->operands = operands;
	return o;
}
Operator * Operator::ternary(Expr* cond, Expr* ifTrue, Expr* ifFalse) {
	Operator * op = new Operator();
	op->op = TERNARY;
	op->op0 = cond;

	std::vector<Expr*> * ops = new std::vector<Expr*>(2);
	ops->push_back(ifTrue);
	ops->push_back(ifFalse);
	op->operands = ops;

	return op;
}
PyObject * Operator::toJson() const {
	PyObject *d = PyDict_New();
	PyDict_SetItemString(d, "op0", op0->toJson());
	PyDict_SetItemString(d, "operator",
			PyUnicode_FromString(OperatorType_toString(op)));

	int arity = OperatorType_arity(op);
	switch (arity) {
	case -1:
	case 3:
		addJsonArrP(d, "operands", *operands);
		break;
	case 1:
		break;
	case 2:
		PyDict_SetItemString(d, "op1", op1->toJson());
		break;
	default:
		throw "Invalid arity of operator";
	}
	return d;
}
